### DISCUSSION TITLE
How do I use Radix with Tailwindcss?

### DISCUSSION THREAD

**Discussion by jan-osch** [radix-ui/primitives#1000] [Help] [answered]
I really like the idea of accessible components! I would like to add some radix primitives to my project. I'm using Next.js and tailwindcss. I would like to add some classes depending on data-state value. I tried using  https://github.com/ecklf/tailwindcss-radix but it doesn't seem to work. I also tried wrapping my components with a forwardRef, like [here](https://www.radix-ui.com/docs/primitives/overview/styling#styling-a-part), but I cannot access the "data-state" of the component. 

I would like to avoid using styled-components or custom CSS classes. How do I get the current state of data-props?

**Comment by jjenzz** [comment]
Which component are you trying to retrieve the `data-state` for? Usually our components can be controlled if you need the state. For example:

```jsx
const [open, setOpen] = React.useState(false);

return (
  <Dialog.Root open={open} onOpenChange={setOpen}>
    <Dialog.Trigger className={open ? 'bg-blue-300' : 'bg-blue-500'}>Open</Dialog.Trigger>
    <Dialog.Content />
  </Dialog.Root>
);
```

**Reply by jan-osch** [reply]
Thanks, @jjenzz, this is exactly what I was looking for. I already adopted the Tabs in my app as a controlled component.

**Reply by vitormarkis** [reply]
i caught myself in similar problem, trying to change the data-state attribute thinking this would reflect wheter the popover is open or not, this state approach is way more useful and makes it so much flexible, thanks for sharing that

**Reply by benoitgrelard** [reply]
Worth noting that now Tailwind supports arbitrary data attributes, so you don't even need to control state to get access to that and styling it.

https://github.com/radix-ui/primitives/discussions/1000#discussioncomment-4888385

**Comment by arekbartnik** [comment]
@jjenzz how can we style using tw utility classes active tab in tabs component?

**Reply by jjenzz** [reply]
Something like this? Obviously the abstraction is totally up to you :)

```jsx
const App = () => {
  const [value, setValue] = React.useState('one');

  const getTriggerClass = (triggerValue) => {
    return triggerValue === value ? 'bg-blue-300' : 'bg-blue-500';
  };

  return (
    <Tabs.Root value={value} onValueChange={setValue}>
      <Tabs.List>
        <Tabs.Trigger value="one" className={getTriggerClass('one')}>
          One
        </Tabs.Trigger>
        <Tabs.Trigger value="two" className={getTriggerClass('two')}>
          Two
        </Tabs.Trigger>
      </Tabs.List>
      <Tabs.Content value="one">One Content</Tabs.Content>
      <Tabs.Content value="two">Two Content</Tabs.Content>
    </Tabs.Root>
  );
};
```

**Comment by ecklf** [comment]
@jan-osch I maintain `tailwindcss-radix`. Would you mind making an issue on what didn't work for you? Happy to get this fixed 🍻

**Reply by mhaidarhanif** [reply]
Thanks @ecklf for the package! I was just looking for it.

**Reply by mhaidarhanif** [reply]
Truly excellent stuff. This is the best way to answer this "how to use Radix with Tailwind CSS".

- https://tailwindcss-radix.vercel.app
- https://github.com/ecklf/tailwindcss-radix

Been using it for [my template project](https://github.com/mhaidarhanif/rewinds) for a while and happy with it.

**Reply by ecklf** [reply]
Glad this package provided value to you @mhaidarhanif <3

**Comment by paulwongx** [comment, marked as answer]
The modern solution seems to be either:
A) Use this plugin by the author of that tailwindcss-radix app above which works (https://www.npmjs.com/package/tailwindcss-radix)
B) Target the data-state directly without a plugin with `data-[state=on]:opacity-40` for example

**Reply by jjenzz** [reply]
yep, tailwind more recently added data attribute support so today i wld ignore all of my suggestions above and use option `B` from @paulwongx here.

**Reply by ecklf** [reply]
Obviously biased opinion since I'm author of option A, but it think you should keep arbitrary values to a minimum unless you are trying to eject out of your design system for edge-cases. It taxes on readability and doesn't give you auto-completion from tailwinds language server.

The recommended way from Tailwind Labs to style HeadlessUI data attributes is to also use variants generated via plugin. See [@headlessui/tailwindcss](https://github.com/tailwindlabs/headlessui/tree/main/packages/@headlessui-tailwindcss).

**Reply by jjenzz** [reply]
sure, whatever works for you. i should have been more specific that i was simply proposing ignoring my suggestions. any solution that allows tailwind to target data attributes is likely going to be a better experience than my original suggestions.

**Comment by ctjlewis** [comment]
The "styles-less" approach is truly bewildering. We have strayed so far from God's grace.

Could literally just ship the CSS with variables and then require consumers to set top-level CSS variables rather than requiring brain surgery to use a component. 

All of web development has been cancelled until further notice. All websites are paused.

**Reply by samism** [reply]
Lmao

**Reply by jjenzz** [reply]
@ctjlewis if radix provided CSS variables they would need to provide variables for every possible thing you might want to style. that would be a maintenance burden for radix and its consumers for no real benefit.

- **what if radix forgets to provide a variable you need?** 
    you have to create a feature request and wait for the team to implement it, **or override to style manually anyway**.
- **what if radix accidentally breaks a variable you're using?** 
    your app is broken so you have to raise a bug report and wait for radix to resolve it, **or override to style manually anyway**.
- **what if your preferred styling approach is BEM?** 
    you add classes to **override and style manually anyway**.

the unstyled approach is magnitudes more flexible.

>  rather than requiring brain surgery to use a component.

i'm curious why adding a `className` with your own styles feels like brain surgery to you?

**Reply by ctjlewis** [reply]
> if radix provided CSS variables they would need to provide variables for every possible thing you might want to style. that would be a maintenance burden for radix and its consumers for no real benefit.

Yes, instead we must provide CSS styles for every possible thing we want to use, which is a huge burden on users! 

I don't know, stuff has gotten so weird now that most people don't seem to mind, but it seems pretty obvious things should come with default styles and simply allow you to override them. We can logically have our cake and eat it too.

Daisy UI is a good example. Chakra etc. These come with default styles.

---

> i'm curious why adding a className with your own styles feels like brain surgery to you?

Relative to just dropping in a component and giving it custom styles if appropriate, it basically is brain surgery.

**Reply by jjenzz** [reply]
> Yes, instead we must provide CSS styles for every possible thing we want to use, which is a huge burden on users!

Radix Primitives are intentionally low-level. their purpose is for people to use them to create their own design systems like Daisy UI or Chakra from them. the libraries you have pointed to are design systems—Radix Primitives is not a design system.

if you are looking for a ready-made design system built from Radix Primitives, here are a couple of open source ones i know of that are available today:

- https://ui.shadcn.com/ 
- https://tamagui.dev/

**Reply by jjenzz** [reply]
@ctjlewis now that it has been announced, i can add https://radix-ui.com/themes to that list 🎉 

Radix Themes is the new design system built on Radix Primitives from the Radix UI team.

**Reply by yp717** [reply]
Full disclosure I'm the maintainer of this package, so just a heads up there. If you're looking to use Tailwind utilities with the Radix palette, `tailwindcss-radix-colors-plugin` could be just what you need. It's a few lines of code to add so takes the burden away from the end user @ctjlewis, and we support themes and theme pinning on DOM subsets (if you want a specific part of the page to stick to a theme) so you can use tailwind like you did before with all the colors in the radix scales.

It might be worth noting we've also made the Radix colors range from 100-1200, like Tailwind's original system, because we found this made it easier to spot them in the markup in a string of utility classes. 

You can download it from [npm](https://www.npmjs.com/package/tailwindcss-radix-colors-plugin?activeTab=versions) and it’s all open source on [GitHub](https://github.com/yp717/tailwindcss-radix-colors-plugin) and we also have a [docs site](https://tailwind-radix-plugin-docs.vercel.app/) with examples + explanations 😄

**Comment by mgibson91** [comment]
You can use radix-ui attributes like:

![image](https://github.com/radix-ui/primitives/assets/3610518/6539d0b1-3b81-49ac-b18b-1e8f8ee01417)

With tailwind like this:

`data-[state=open]:bg-green-500`

Incredibly useful - thanks Radix team!

**Comment by junwen-k** [comment]
To anyone who is reading this, recently I've just released a [plugin](https://github.com/junwen-k/tailwindcss-radix-ui-primitives) that simplifies styling for Radix UI using Tailwind CSS. It works similarly with [tailwindcss-radix](https://github.com/ecklf/tailwindcss-radix) but the difference is that it injects Radix UI common data attribute selectors directly to your [`theme.data`](https://tailwindcss.com/docs/hover-focus-and-other-states#data-attributes).

You can then write Radix's state variants with the benefit of IntelliSense. For example:

```diff
- class="data-[state=open]:..."
+ class="data-state-open:..."
```

Personally I prefer this way because it feels more "idiomatic" as Tailwind CSS recently added support for [data-attribute selectors](https://tailwindcss.com/blog/tailwindcss-v3-2#data-attribute-variants).
Additionally, CSS variables like `w-radix-accordion-content-height` are also supported.

The plugin is heavily inspired by [tailwindcss-radix](https://github.com/ecklf/tailwindcss-radix), thanks @ecklf!

**Comment by oussamachah2020** [comment]
i just want to change button background LOL

**Comment by FlatMapIO** [comment]
I used different approaches to import radix-theme in SolidJS and Tailwind. I prefer the method of customizing based on copy-paste modification, so I provided a solution with CSS modules + CVA.

https://github.com/FlatMapIO/radix-theme-solid-start?tab=readme-ov-file
